enhancements/network/static-ip-addresses-vsphere.md

```diff
@@ -0,0 +1,537 @@
+	// to this device.
+	// Required when DHCP4 and DHCP6 are both Disabled.
+	// + Validation is applied via a patch, we validate the format as either ipv4 or ipv6
+	// +optional
+	IPAddrs []string `json:"ipAddrs,omitempty"`
+
+	// nameservers is a list of IPv4 and/or IPv6 addresses used as DNS
+	// nameservers.
+	// Please note that Linux allows only three nameservers (https://linux.die.net/man/5/resolv.conf).
+	// +optional
+	Nameservers []string `json:"nameservers,omitempty"`
+  
+  // addressesFromPools is a list of address pools which will fulfill requests for static IPs
+  AddressesFromPools []AddressesFromPool `json:"addressesFromPool,omitempty"`
+}
+
+~~~
+
+Example of a platform spec configured to provide static IPs for the bootstrap, control plane, and compute nodes:
+~~~yaml
+platform:
+  vsphere:
+    hosts:
+    - role: bootstrap
+      networkDevice:
+        ipAddrs:
+        - 192.168.101.240/24
+        gateway4: 192.168.101.1
+        nameservers:
+        - 192.168.101.2
+    - role: control-plane
+      failureDomain: us-east-1a
+      networkDevice:
+        ipAddrs:
+        - 192.168.101.241/24
+        gateway4: 192.168.101.1
+        nameservers:
+        - 192.168.101.2
+    - role: control-plane
+      failureDomain: us-east-1b
+      networkDevice:
+        ipAddrs:
+        - 192.168.101.242/24
+        gateway4: 192.168.101.1
+        nameservers:
+        - 192.168.101.2
+    - role: control-plane
+      failureDomain: us-east-1c
+      networkDevice:
+        ipAddrs:
+        - 192.168.101.243/24
+        gateway4: 192.168.101.1
+        nameservers:
+        - 192.168.101.2
+    - role: compute
+      networkDevice:
+        ipAddrs:
+        - 192.168.101.244/24
+        gateway4: 192.168.101.1
+        nameservers:
+        - 192.168.101.2
+    - role: compute
+      networkDevice:
+        ipAddrs:
+        - 192.168.101.245/24
+        gateway4: 192.168.101.1
+        nameservers:
+        - 192.168.101.2
+    - role: compute
+      networkDevice:
+        ipAddrs:
+        - 192.168.101.246/24
+        gateway4: 192.168.101.1
+        nameservers:
+        - 192.168.101.2
+~~~
+2. Add validation for the modified/added fields in the platform specification.
+3. For compute nodes, produce machine manifests with associated network device configuration.  
+
+Example of `machine` configured with network device configuration
+~~~yaml
+apiVersion: machine.openshift.io/v1beta1
+kind: Machine
+metadata:
+  name: test-compute-1
+spec:
+  metadata: {}
+  providerSpec:
+    value:
+      numCoresPerSocket: 2
+      diskGiB: 60
+      snapshot: ''
+      userDataSecret:
+        name: worker-user-data
+      memoryMiB: 8192
+      credentialsSecret:
+        name: vsphere-cloud-credentials
+      network:
+        devices:
+          - networkName: lab
+            ipAddrs:
+            - 192.168.101.244/24
+            gateway4: 192.168.101.1
+            nameservers:
+            - 192.168.101.2            
```

Comment:
What would this look like when CPMS tries to adopt it?

Reply:
The vision for CPMS is that the IP address would not be specified statically in the machine spec. Rather, CPMS leverage an IP pool and rely upon `IPAddressClaim` to receive an IP address for the machine.

Reply:
Why not have 1 path that covers all use cases, I worry that having multiple paths is confusing, is this what upstream did?

Reply:
One of the benefits of keeping this is that the external IPAM controller isn't required during install.  We could eliminate the ability to define the static IPs in the spec and require the installation to be done on a compact cluster.  At present, even during an installation, control plane nodes don't embed static IP configuration.

Reply:
Do we have, anywhere in here, an explanation of how CPMS will cover masters? Has any POC been done?

Reply:
I will update this document with CPMS journey / details shortly